src/Storages/MergeTree/MergeTreeRangeReader.cpp

```diff
@@ -866,7 +925,7 @@ void MergeTreeRangeReader::executePrewhereActionsAndFilterColumns(ReadResult & r
     if (result.totalRowsPerGranule() == 0)
         result.setFilterConstFalse();
     /// If we need to filter in PREWHERE
-    else if (prewhere->need_filter || result.need_filter)
+    else if (prewhere_info_list->back().need_filter || result.need_filter)
```

Comment:
Note that `need_filter` flag is ignored for other cases. It works now because this flag is true for row-level security.